### ISSUE TITLE
jsHint validthis equivalent

### ISSUE THREAD

**Issue by mzgol** [eslint/eslint#1636]
jsHint has a `validthis` rule, enabled by default, that errors when a regular function (not a method) uses `this` somewhere in the body. It can be useful.

If one is certain a specific function is only used as a method, it's always possible to overwrite the rule for this function:
```js
function getResult() {
    /* jshint validthis:true */
    return this.getResult();
}
// ... later ...
getResult.call(object);
```

Could sth like that be added to ESLint? I didn't see any issue mentioning this rule.

EDIT: simplified the example.

**Comment by nzakas** [comment]
This would be a bit tricky to implement since it's a rule that can only be applied inside of a function. What we could do instead is create a rule that flags the use of `this` in strict mode functions, and then you could disable where appropriate:

```js
function getResult() {
    "use strict";

    /*eslint-disable strict-this*/
    return this.getResult();
    /*eslint-enable strict-this*/
}
```

Would that work?

**Comment by mzgol** [comment]
That would work but with a few conditions:

1. All our code is in IIFEs covering whole files and applying strict mode globally per file; this would have to work with the following code as well:
    ```js

    (function () {
        "use strict";

        function getResult() {
            /*eslint-disable strict-this*/
            return this.getResult();
            /*eslint-enable strict-this*/
        }
    })();
    ```
2. Methods should be exempt from the rule, i.e. the following:
    ```js

    var o = {
        a: 2,
        f: function () {
            return this.a;
        },
    };
    ```
should lint correctly without any additional comments.

**Comment by nzakas** [comment]
Yup, both of those are possible.

**Comment by doberkofler** [comment]
+1

**Comment by gyandeeps** [comment]
@ilyavolodin @nzakas I think an agreement was reached here but we are missing the appropiate labels here.

**Comment by nzakas** [comment]
Updated

**Comment by gyandeeps** [comment]
Would PR https://github.com/eslint/eslint/pull/2857 solve this issue?

**Comment by mzgol** [comment]
It seems it would. :)

**Comment by gyandeeps** [comment]
closing as per the above discussion.

**Comment by mysticatea** [comment]
Sorry, I had been missing this issue.

**Comment by mysticatea** [comment]
FYI: When there is the `@this` tag in JSDoc comment, `this`s in the function are not warned, it's similar to `/* jshint validthis:true */`.
